File: GRL/05/a.py
from collections import deque

MAX = 100000
INF = float('inf')


class Edge:

    def __init__(self, t, w):
        self.t = t
        self.w = w

G = [[] for i in xrange(MAX)]
n = 0
d = []


def bfs(s):
    global d
    d = [INF] * n
    q = deque()
    q.append(s)
    d[s] = 0
    while len(q):
        u = q.popleft()
        for i in xrange(len(G[u])):
            e = G[u][i]
            if d[e.t] == INF:
                d[e.t] = d[u] + e.w
                q.append(e.t)


def solve():
    bfs(0)
    maxv = 0
    tgt = 0
    for i in xrange(n):
        if d[i] == INF:
            continue
        if maxv < d[i]:
            maxv = d[i]
            tgt = i

    bfs(tgt)
    maxv = 0
    for i in xrange(n):
        if d[i] == INF:
            continue
        maxv = max(maxv, d[i])

    print maxv


def main():
    global n
    n = input()
    for i in xrange(n - 1):
        s, t, w = map(int, raw_input().split())
        G[s].append(Edge(t, w))
        G[t].append(Edge(s, w))
    solve()

if __name__ == '__main__':
    main()

File: GRL/05/a.cpp
#include<bits/stdc++.h>
using namespace std;
#define MAX 100000
#define INF (1 << 30)

class Edge {
public:
    int t, w;
    Edge(){}
    Edge(int t, int w): t(t), w(w) {}
};

vector<Edge> G[MAX];
int n, d[MAX];

void bfs(int s) {
    for (int i = 0; i < n; i++) d[i] = INF;
    queue<int> q;
    q.push(s);
    d[s] = 0;
    int u;
    while (!q.empty()) {
        u = q.front(); q.pop();
        for (int i = 0; i < G[u].size(); i++) {
            Edge e = G[u][i];
            if (d[e.t] == INF) {
                d[e.t] = d[u] + e.w;
                q.push(e.t);
            }
        }
    }
}

void solve() {
    bfs(0);
    int maxv = 0;
    int tgt = 0;
    for (int i = 0; i < n; i++) {
        if (d[i] == INF) continue;
        if (maxv < d[i]) {
            maxv = d[i];
            tgt = i;
        }
    }

    bfs(tgt);
    maxv = 0;
    for (int i = 0; i < n; i++) {
        if (d[i] == INF) continue;
        maxv = max(maxv, d[i]);
    }

    cout << maxv << endl;
}

int main() {
    int s, t, w;
    cin >> n;
    for (int i = 0; i < n-1; i++) {
        cin >> s >> t >> w;
        G[s].push_back(Edge(t, w));
        G[t].push_back(Edge(s, w));
    }
    
    solve();
    return 0;
}
